feat: add server timing to responses

SgtPooki opened the pull request:
## Title

<!---
The title of the PR will be the commit message of the merge commit, so please make sure it is descriptive enough.
We utilize the Conventional Commits specification for our commit messages. See <https://www.conventionalcommits.org/en/v1.0.0/#specification> for more information.
The commit tag types can be of one of the following: feat, fix, deps, refactor, chore, docs. See <https://github.com/ipfs/helia/blob/main/.github/workflows/main.yml#L184-L192>
The title must also be fewer than 72 characters long or it will fail the Semantic PR check. See <https://github.com/ipfs/helia/blob/main/.github/workflows/semantic-pull-request.yml>
--->
feat: add server timing to responses

## Description

<!--
Please write a summary of your changes and why you made them.
Please include any relevant issues in here, for example:
Related https://github.com/ipfs/helia/issues/ABCD.
Fixes https://github.com/ipfs/helia/issues/XYZ.
-->

This PR adds server timing headers to responses by wrapping all responses returned by `fetch` in a `Response` object that includes the server timing headers.

The headers are generated by wrapping async functions with a `serverTiming` utility function that records the time taken to execute the function, and then adding those headers to an array on the VerifiedFetch class.

## Notes & open questions

<!--
Any notes, remarks or open questions you have to make about the PR which don't need to go into the final commit message.
-->

We need to work out exactly what events we need to time, and I'd also like to clean up the `serverTiming` utility function usage so that it's easier to add new timings without having to handle the error and success cases manually for each one.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation if necessary (this includes comments as well)
- [ ] I have added tests that prove my fix is effective or that my feature works

SgtPooki commented:
with https://github.com/ipfs/service-worker-gateway/pull/531

we get:

<img width="2137" alt="image" src="https://github.com/user-attachments/assets/7dcffb36-30c4-4aa7-8acf-16f221375f00" />

SgtPooki reviewed (commented):
self review

SgtPooki reviewed (commented):
self review of added test

2color reviewed (commented):
This is pretty neat. Implementation looks good to me.

What do you think about making this optional per request? I can imagine that in node.js this isn't very useful and just makes the response object larger. 

> We need to work out exactly what events we need to time

At the very least we'd want: 
- IPNS/DNSLink resolution (if it's an ipns url) 
- content routing, i.e. finding providers.

SgtPooki commented:
@2color added the option to include server timing headers, defaulting to false on construction but overridable per request. Also added docs and updated readme

2color commented on packages/verified-fetch/src/index.ts:
```diff
@@ -604,6 +604,14 @@
  *
+ * #### Server Timing headers
+ *
+ * By default, we do not include Server Timing headers in responses. If you want to include them, you can pass an
```
```suggestion
 * By default, we do not include [Server Timing](https://developer.mozilla.org/en-US/docs/Web/API/Performance_API/Server_timing) headers in responses. If you want to include them, you can pass an
```

2color commented on packages/verified-fetch/src/index.ts:
```diff
@@ -750,6 +758,14 @@ export interface CreateVerifiedFetchOptions {
+   * @default false
+   * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Server-Timing
+   */
+  includeServerTiming?: boolean
```
Since it's typed as a boolean, any thoughts on making this a bit shorter?
```suggestion
  serverTiming?: boolean
```

SgtPooki replied:
I don't have a strong preference, but I prefer a more explicit name since `serverTiming` doesn't communicate its purpose. Also, verb-prefixed names for booleans are more idiomatic. What do you think about `withServerTiming` instead?

Option names should communicate their intent and effect rather than aim for brevity. More explicit names lead to a better dev understanding, with and without reading the docs. Also, the options are only passed once.

2color replied:
Fair enough! I agree. 

`withServerTiming` works too and is shorter.

2color commented on packages/verified-fetch/src/verified-fetch.ts:
```diff
@@ -452,6 +460,26 @@ export class VerifiedFetch {
     [identity.code]: this.handleRaw
   }
 
+  private async handleServerTiming<T> (name: string, description: string, fn: () => Promise<T>): Promise<T> {
```
Can we check `this.includeServerTiming`  here to encapsulate most of the logic here and avoid the measuring and assigning of `this.serverTimingHeaders`? 

That would also allow simplifying `responseWithServerTiming`  and renaming to something a bit more generic, e.g. `augmentResponse` or `handleResponse`

SgtPooki replied:
yea let me peek at this

SgtPooki replied:
so, it just so happened that switching the server timing handling to handleServerTiming required me to pipe things through a lot more, so I'm moving forward with this and adding the "IPNS/DNSLink resolution (if it's an ipns url)" you previously mentioned.

I didn't originally like it in `handleServerTiming` because it required plumbing the request-level "withServerTiming" into more places, but it makes more sense for multiple reasons. 

it also prevents the extra work of wrapping the timed functions.. so it's a win. thanks for the push

SgtPooki commented:
@2color all requested changes are in, but the code is a little bit sloppier now because of how parse-resource & parse-url-string are currently written.

2color commented on packages/verified-fetch/src/utils/parse-url-string.ts:
```diff
@@ -207,7 +225,19 @@ export async function parseUrlString ({ urlString, ipns, logger }: ParseUrlStrin
+            if (resolveResultWithServerTiming.error != null) {
+              throw resolveResultWithServerTiming.error
+            }
+            resolveResult = resolveResultWithServerTiming.result ?? undefined
```
```suggestion
            resolveResult = resolveResultWithServerTiming.result
```

SgtPooki replied:
this was needed for typescript and then I updated some code.. just checked if we can remove the null coalescing to undefined and it seems like we can.. just pushed up the change

2color reviewed (commented):
LGTM. I still think it would be really useful to add a duration for finding providers from each router/block-broker. 

Since that would probably require us to add progress hooks to the helia routers, we can handle this in a follow up PR. 

Not directly related, but more insight into routing would help surface more nuanced errors: no providers vs. failure to connect to providers, to the user.

SgtPooki commented:
CI seems to be failing when attempting to install Kubo on windows.. i'm looking into it and will try clearing package-lock and updating Kubo

SgtPooki commented:
> CI seems to be failing when attempting to install Kubo on windows.. i'm looking into it and will try clearing package-lock and updating Kubo

Nikolai is looking into the collab cluster
